File: windows/src/BetterScreenshot.App/Capture/CaptureCoordinator.cs
using System.Diagnostics;
using System.IO;
using System.Windows.Media.Imaging;
using BetterScreenshot.App.Editor;
using BetterScreenshot.App.History;
using BetterScreenshot.App.Overlays;
using BetterScreenshot.App.Recording;
using BetterScreenshot.App.Tray;
using BetterScreenshot.Capture;
using BetterScreenshot.Core;
using BetterScreenshot.Platform;

namespace BetterScreenshot.App.Capture;

/// <summary>
/// Orchestrates the screenshot flow: selection/picker overlays → capture (via the Platform layer) → route by the
/// after-capture setting → save/copy/Quick Access card, plus Capture Text (region OCR), pins, history, and the
/// recording toggle.
/// </summary>
public sealed class CaptureCoordinator : IAppCommands
{
    private readonly SettingsStore _settings;
    private readonly Action _quit;
    private readonly HistoryService _history;
    private readonly SelectionOverlayController _selection = new();
    private readonly QuickAccessStackController _stack = new();
    private readonly PinPanelController _pins = new();
    private readonly WindowPickerController _picker = new();
    private readonly RecordingCoordinator _recording;
    private HistoryWindow? _historyWindow;

    public CaptureCoordinator(SettingsStore settings, Action quit)
    {
        _settings = settings;
        _quit = quit;
        _history = HistoryService.ForSettings(settings);
        _recording = new RecordingCoordinator(
            settings,
            (recording, elapsed) => OnRecordingStateChanged?.Invoke(recording, elapsed),
            (active, paused) => OnRecordingPauseChanged?.Invoke(active, paused),
            OnRecordingFinished);
    }

    /// <summary>Set by the app to show the settings window (which needs the hotkey controller too).</summary>
    public Action? OnOpenSettings { get; set; }

    /// <summary>Set by the app to reflect recording state (icon + elapsed timer) in the tray.</summary>
    public Action<bool, string?>? OnRecordingStateChanged { get; set; }

    /// <summary>Set by the app to reflect pause state (active?, paused?) on the tray Pause/Resume menu item.</summary>
    public Action<bool, bool>? OnRecordingPauseChanged { get; set; }

    public void CaptureFullscreen()
    {
        var monitor = Screens.Primary();
        Handle(ScreenCapture.CaptureDisplay(monitor));
    }

    public void CaptureWindow()
    {
        // Freeze first, while the app you were using still has focus — the picker overlay is what makes it lose it.
        var frozen = Freeze();
        _picker.Present(pick =>
        {
            if (pick is { } p && p.Hwnd != IntPtr.Zero) Handle(p.Frozen ?? ScreenCapture.CaptureWindow(p.Hwnd));
        }, frozen);
    }

    public void CaptureArea()
    {
        _selection.Present(_settings.Capture.FreezeScreen, selection =>
        {
            if (selection is { } s) Handle(Pixels(s));
        });
    }

    /// <summary>Capture Text (OCR + QR): drag a region; the recognized text — or a QR code's payload,
    /// which wins — lands on the clipboard. HUD confirms.</summary>
    public void CaptureText()
    {
        _selection.Present(_settings.Capture.FreezeScreen, selection =>
        {
            if (selection is { } s) _ = CaptureTextAsync(s);
        });
    }

    /// <summary>A still of every screen when "Freeze the screen" is on, else null (overlays stay see-through
    /// and the capture is taken live).</summary>
    private FrozenScreen? Freeze() => _settings.Capture.FreezeScreen ? FrozenScreen.Capture() : null;

    /// <summary>The selected pixels: cropped out of the frozen still when there is one, else captured live.</summary>
    private static BitmapSource Pixels(AreaSelection selection) =>
        selection.Frozen ?? ScreenCapture.CaptureRegion(selection.Region);

    private async Task CaptureTextAsync(AreaSelection selection)
    {
        try
        {
            var image = Pixels(selection);
            var result = await TextRecognizerService.RecognizeAsync(image);
            if (result.ClipboardString is { } text) ClipboardService.SetText(text);
            HudController.Show(result.HudMessage);
        }
        catch
        {
            // Never crash the app on a failed recognition.
            HudController.Show("Capture Text failed");
        }
    }

    private void Handle(BitmapSource image)
    {
        var (copy, save, overlay) = CaptureRouter.Decide(_settings.Capture.AfterCapture);
        if (copy) Copy(image);
        if (save) Save(image);
        // Remember every non-ephemeral capture (saved or shown) in history; copy-only captures stay transient.
        Guid? historyId = (save || overlay) ? _history.RecordScreenshot(image) : null;
        if (overlay) ShowOverlayCard(image, historyId);
    }

    private void ShowOverlayCard(BitmapSource image, Guid? historyId = null)
    {
        string dragFile = ImageIo.WriteTempPng(image, "quickaccess.png");
        var actions = new QuickAccessActions
        {
            OnCopy = () => Copy(image),
            OnSave = () => Save(image),
            OnPin = () => PinImage(image),
            OnEdit = () => Annotate(image),
        };
        _stack.Present(image, QuickAccessKind.Screenshot, actions, MapCorner(_settings.Capture.OverlayCorner),
            dragFile, _settings.Capture.OverlayAutoDismissSeconds, reason =>
            {
                OnCardDismissed(historyId, reason);
                // The temp PNG only backs drag-to-export. Keep it alive for PayloadLifetime (5 min) after the card
                // goes away so an in-flight drop into another app can still read the file, then auto-delete it. The
                // screenshot itself is preserved separately in History (its own %APPDATA% copy), so this never loses
                // the capture — it only cleans up the throwaway drag file.
                TempFiles.ScheduleDeleteContainingDir(dragFile, TempFiles.PayloadLifetime);
            });
    }

    /// <summary>Only ✕-closed / evicted cards are restorable; deliberate actions (save/edit/pin) are not.</summary>
    private void OnCardDismissed(Guid? historyId, DismissReason reason)
    {
        if (historyId is { } id && reason is DismissReason.Closed or DismissReason.Evicted)
            _history.NoteClosed(id);
    }

    private static Corner MapCorner(SettingsOverlayCorner corner) => corner switch
    {
        SettingsOverlayCorner.TopLeft => Corner.TopLeft,
        SettingsOverlayCorner.TopRight => Corner.TopRight,
        SettingsOverlayCorner.BottomLeft => Corner.BottomLeft,
        _ => Corner.BottomRight,
    };

    private void Save(BitmapSource image)
    {
        bool jpg = _settings.Capture.Format == SettingsImageFormat.Jpg;
        string ext = jpg ? "jpg" : "png";
        string path = Path.Combine(_settings.SaveDirectory, FileNamer.Name(DateTime.Now, ext));
        if (jpg) ImageIo.SaveJpg(image, path); else ImageIo.SavePng(image, path);
    }

    private static void Copy(BitmapSource image) => ClipboardService.SetImage(image);

    public void PinFromClipboard()
    {
        if (System.Windows.Clipboard.ContainsImage())
        {
            var image = System.Windows.Clipboard.GetImage();
            if (image != null) PinImage(image);
        }
    }

    private void PinImage(BitmapSource image)
    {
        var style = new PinStyle(_settings.Capture.PinCornerRadius, _settings.Capture.PinShadow);
        var actions = new PinActions(() => Copy(image), () => Save(image));
        _pins.Pin(image, style, actions);
    }

    /// <summary>Opens the annotation editor on the image, wiring copy/save/stack and sticky-style persistence.</summary>
    private void Annotate(BitmapSource image)
    {
        var editor = new EditorWindow(image, _settings.EditorStyle)
        {
            OnCopy = Copy,
            OnSave = Save,
            OnAddToStack = KeepInStack,
            StyleChanged = style => { _settings.EditorStyle = style; _settings.Save(); },
        };
        editor.Show();
    }

    /// <summary>Editor "Stack" button: record the flattened edit in history and re-enter the Quick Access flow.</summary>
    private void KeepInStack(BitmapSource image)
    {
        var id = _history.RecordScreenshot(image);
        ShowOverlayCard(image, id);
    }

    /// <summary>Restore Recently Closed: bring back the newest ✕-closed card that still exists in history.</summary>
    public void RestoreRecentlyClosed()
    {
        var entry = _history.PopRestorable();
        if (entry is null) return;
        var image = _history.LoadImage(entry);
        if (image != null) ShowOverlayCard(image, entry.Id);
    }

    /// <summary>Shows the capture-history browser (a single reused window), reloading its grid each time.</summary>
    public void OpenHistory()
    {
        if (_historyWindow is null)
        {
            _historyWindow = new HistoryWindow(_history, new HistoryWindowActions(Annotate, PinImage));
            _historyWindow.Closed += (_, _) => _historyWindow = null;
            _historyWindow.Show();
        }
        else
        {
            _historyWindow.Activate();
        }
    }

    /// <summary>Start/stop a screen recording (record strip → target picking).</summary>
    public void ToggleRecording() => _recording.Toggle();

    /// <summary>Pause or resume the active recording (gapless via segment+concat).</summary>
    public void PauseResumeRecording() => _recording.PauseResume();

    /// <summary>A finished recording: record it in history and show a Quick Access recording card.</summary>
    private void OnRecordingFinished(string path, BitmapSource thumbnail)
    {
        Guid? id = _history.RecordRecording(path, thumbnail);
        ShowRecordingCard(path, thumbnail, id);
    }

    private void ShowRecordingCard(string path, BitmapSource thumbnail, Guid? historyId)
    {
        var actions = new QuickAccessActions
        {
            OnCopy = () => ClipboardService.SetFile(path),
            OnOpen = () => OpenFile(path),
            OnReveal = () => RevealFile(path),
        };
        // Recording cards drag the real saved file (never a temp copy) — so it is NOT scheduled for deletion.
        _stack.Present(thumbnail, QuickAccessKind.Recording, actions, MapCorner(_settings.Capture.OverlayCorner),
            path, _settings.Capture.OverlayAutoDismissSeconds, reason => OnCardDismissed(historyId, reason));
    }

    private static void OpenFile(string path)
    {
        try { Process.Start(new ProcessStartInfo(path) { UseShellExecute = true }); }
        catch { /* best-effort: no default handler */ }
    }

    private static void RevealFile(string path)
    {
        try { Process.Start(new ProcessStartInfo("explorer.exe", HistoryService.ExplorerSelectArgs(path)) { UseShellExecute = true }); }
        catch { /* best-effort */ }
    }

    /// <summary>Called on app exit: best-effort finalize an in-progress recording so it isn't lost.</summary>
    public void StopRecordingForExit() => _recording.StopForExit();

    public void OpenSettings() => OnOpenSettings?.Invoke();
    public void Quit() => _quit();
}

File: App/Capture/CaptureCoordinator.swift
import AppKit
import ScreenCaptureKit
import CaptureKit
import OverlayKit
import EditorKit

@MainActor
final class CaptureCoordinator {
    private let service = CaptureService()
    private let settings: SettingsStore
    private let overlay = SelectionOverlayController()
    // Shared with RecordingCoordinator so screenshot and recording overlays
    // stack together at the corner instead of overlapping.
    private let quickAccess: QuickAccessStackController
    private let hud = HUDController()
    private let pins = PinPanelController()

    /// Filled in by Plan 3 to present the annotation editor. Nil = stub.
    var editorPresenter: ((CGImage) -> Void)?

    /// Set by the app delegate; presents the one-button permission setup window.
    var presentSetup: (() -> Void)?

    /// Set by the app delegate; nil until then (history silently skipped).
    var history: HistoryService?

    private var editorController: EditorWindowController?

    /// The last app that was frontmost before one of our overlays took focus,
    /// so focus can be handed back after a capture. Outlives a single capture
    /// on purpose — see `rememberFrontmostApp()`.
    private var previousApp: NSRunningApplication?

    func presentEditor(_ image: CGImage) {
        let controller = EditorWindowController(image: image, defaultStyle: settings.editorStyle,
                                                recentColors: settings.editorRecentColors)
        controller.onCopy = { [weak self] img in self?.copy(img) }
        controller.onSave = { [weak self] img in self?.save(img) }
        controller.onAddToStack = { [weak self] img in self?.keepInStack(img) }
        controller.onStyleChanged = { [weak self] style in
            self?.settings.editorStyle = style
            self?.settings.persistEditorStyle()
        }
        controller.onRecentColorsChanged = { [weak self] colors in
            self?.settings.editorRecentColors = colors
            self?.settings.persistEditorRecentColors()
        }
        editorController = controller
        controller.showWindow(nil)
        controller.window?.makeKeyAndOrderFront(nil)
        NSApp.activate(ignoringOtherApps: true)
    }

    /// Drops an edited image into the bottom-right Quick Access stack, treating
    /// it like a fresh capture: it is recorded to history and shown with the
    /// normal Copy/Edit/Pin/Save/drag actions.
    func keepInStack(_ image: CGImage) {
        let historyID = history?.recordScreenshot(image)
        presentOverlay(image, sourceRect: nil, historyID: historyID)
    }

    init(settings: SettingsStore, quickAccess: QuickAccessStackController) {
        self.settings = settings
        self.quickAccess = quickAccess
    }

    func captureArea() {
        rememberFrontmostApp()
        guard ensurePermission() else { return }
        overlay.present { [weak self] result in
            guard let self else { return }
            guard let result else { self.restoreFrontmostApp(); return }
            Task { await self.run(.area(rect: result.globalRect, displayID: result.displayID),
                                  sourceRect: result.globalRect) }
        }
    }

    func captureFullscreen() {
        rememberFrontmostApp()
        guard ensurePermission() else { return }
        Task { await run(.fullscreen(displayID: CGMainDisplayID())) }
    }

    func captureFrontWindow() {
        rememberFrontmostApp()
        guard ensurePermission() else { return }
        Task { if let id = await frontmostWindowID() { await run(.window(windowID: id)) } }
    }

    /// Capture Text (OCR + QR): drag a region; the recognized text — or a QR
    /// code's payload, which wins — lands on the clipboard. HUD confirms.
    func captureText() {
        rememberFrontmostApp()
        guard ensurePermission() else { return }
        // Load Vision's text model while the user drags — cold start is 0.5–1s.
        Task.detached(priority: .userInitiated) { TextRecognizer.warmUp() }
        overlay.present { [weak self] result in
            guard let self else { return }
            guard let result else { self.restoreFrontmostApp(); return }
            Task { await self.runCaptureText(result) }
        }
    }

    private func runCaptureText(_ result: SelectionResult) async {
        do {
            let image = try await service.capture(
                .area(rect: result.globalRect, displayID: result.displayID))
            // Vision's perform() blocks — keep it off the main actor.
            let pointWidth = result.globalRect.width
            let recognition = try await Task.detached {
                try TextRecognizer.recognize(in: image, pointWidth: pointWidth)
            }.value
            if let payload = recognition.clipboardString {
                NSPasteboard.general.clearContents()
                NSPasteboard.general.setString(payload, forType: .string)
            }
            hud.show(recognition.hudMessage, symbol: "text.viewfinder", on: screen(for: result.displayID))
        } catch {
            NSLog("Capture Text failed: \(error)")
            hud.show("Capture Text failed", symbol: "exclamationmark.triangle", on: screen(for: result.displayID))
        }
        // Focus goes back last, so the recognized text can be pasted straight
        // into the app the user was already in.
        restoreFrontmostApp()
    }

    private func screen(for displayID: CGDirectDisplayID) -> NSScreen? {
        NSScreen.screens.first {
            ($0.deviceDescription[NSDeviceDescriptionKey("NSScreenNumber")] as? NSNumber)?
                .uint32Value == displayID
        } ?? NSScreen.main
    }

    private func run(_ target: CaptureTarget, sourceRect: CGRect? = nil) async {
        do {
            let image = try await service.capture(target)
            handle(image, sourceRect: sourceRect)
        } catch {
            NSLog("Capture failed: \(error)")
            hud.show("Capture failed", symbol: "exclamationmark.triangle")
            restoreFrontmostApp()
        }
    }

    private func handle(_ image: CGImage, sourceRect: CGRect?) {
        // Silent bookkeeping first, so even copy-only captures are recoverable.
        let historyID = history?.recordScreenshot(image)
        if settings.settings.playSound { CaptureSound.play() }
        switch settings.settings.afterCapture {
        case .copyOnly:    copy(image)
        case .saveOnly:    save(image)
        case .copyAndSave: copy(image); save(image)
        case .showOverlay: presentOverlay(image, sourceRect: sourceRect, historyID: historyID)
        }
        // The Quick Access card is a .nonactivatingPanel, so handing focus back
        // here leaves it on screen and clickable.
        restoreFrontmostApp()
    }

    private func presentOverlay(_ image: CGImage, sourceRect: CGRect?, historyID: UUID?) {
        let nsImage = NSImage(cgImage: image,
                              size: NSSize(width: image.width, height: image.height))
        guard let screen = NSScreen.main else { copy(image); save(image); return }
        let actions = QuickAccessActions(
            onCopy: { [weak self] in self?.copy(image); self?.hud.show("Copied", symbol: "doc.on.doc") },
            // The overlay's download button always lands in the macOS screenshot folder.
            onSave: { [weak self] in self?.save(image, to: SettingsStore.systemScreenshotLocation()) },
            onAnnotate: { [weak self] in self?.annotate(image) },
            fileURLForDrag: { TempImageWriter.writePNG(image, fileName: FileNamer.fileName(for: Date(), ext: "png")) })
        let corner = settings.settings.overlayCorner
        // visibleFrame excludes the Dock and menu bar, so the overlay sits above
        // the Dock instead of being tucked into the very bottom corner behind it.
        let frame = screen.visibleFrame
        quickAccess.present(image: nsImage, actions: actions,
                            autoDismissSeconds: settings.settings.overlayAutoDismissSeconds,
                            corner: corner, screenFrame: frame, margin: 24,
                            onDismissed: { [weak self] reason in
            // ✕-close and eviction are "accidental" — deliberate actions aren't restorable.
            if reason == .closed || reason == .evicted {
                self?.history?.noteOverlayClosed(historyID: historyID)
            }
        })
    }

    /// Re-presents a Quick Access card for a history entry (Restore Recently Closed).
    func presentOverlayFromHistory(_ image: CGImage, historyID: UUID) {
        presentOverlay(image, sourceRect: nil, historyID: historyID)
    }

    /// Plan 3 replaces the stub body via `editorPresenter`.
    func annotate(_ image: CGImage) {
        if let present = editorPresenter { present(image) }
        else { NSLog("Annotate requested — editor arrives in Plan 3") }
    }

    /// Pins the image as a floating panel — at its original on-screen location
    /// when known, else centered on the main screen.
    func pin(_ image: CGImage, near sourceRect: CGRect? = nil) {
        guard image.width > 0, image.height > 0 else { return }
        let screen = sourceRect.flatMap { r in NSScreen.screens.first { $0.frame.intersects(r) } }
            ?? NSScreen.main
        guard let screen else { return }
        let nsImage = NSImage(cgImage: image,
                              size: NSSize(width: image.width, height: image.height))
        let style = PinStyle(cornerRadius: CGFloat(settings.settings.pinCornerRadius),
                             shadow: settings.settings.pinShadow)
        let actions = PinActions(
            onCopy: { [weak self] in
                self?.copy(image)
                // Re-resolve at click time: the original display may be gone.
                let liveScreen = NSScreen.screens.first { $0 === screen } ?? NSScreen.main
                self?.hud.show("Copied", symbol: "doc.on.doc", on: liveScreen)
            },
            onSave: { [weak self] in self?.save(image) })
        pins.pin(image: nsImage,
                 pixelSize: CGSize(width: image.width, height: image.height),
                 sourceRect: sourceRect, on: screen, style: style, actions: actions)
    }

    func pinFromClipboard() {
        guard let ns = NSPasteboard.general.readObjects(forClasses: [NSImage.self],
                                                        options: nil)?.first as? NSImage,
              let cg = ns.cgImage(forProposedRect: nil, context: nil, hints: nil) else {
            hud.show("No image on clipboard", symbol: "exclamationmark.triangle", on: NSScreen.main)
            return
        }
        pin(cg)
    }

    var clipboardHasImage: Bool {
        NSPasteboard.general.canReadObject(forClasses: [NSImage.self], options: nil)
    }

    private func copy(_ image: CGImage) {
        let rep = NSBitmapImageRep(cgImage: image)
        let nsImage = NSImage(); nsImage.addRepresentation(rep)
        NSPasteboard.general.clearContents()
        // Image data first (so image-targets paste the picture), plus a real PNG
        // file so file-targets — a terminal/Claude Code you paste into — get a
        // usable path. TempFileService sweeps the file once it passes the
        // "Keep cached files for" window.
        var objects: [NSPasteboardWriting] = [nsImage]
        if let url = TempImageWriter.writePNG(image,
                                              fileName: FileNamer.fileName(for: Date(), ext: "png")) {
            objects.append(url as NSURL)
        }
        NSPasteboard.general.writeObjects(objects)
    }

    private func save(_ image: CGImage, to directory: URL? = nil) {
        let dir = directory ?? settings.saveDirectory
        let isPNG = settings.settings.format == .png
        let format: ImageFormat = isPNG ? .png : .jpg(quality: 0.9)
        guard let data = ImageEncoder.encode(image, as: format) else {
            hud.show("Couldn't save — image encoding failed", symbol: "exclamationmark.triangle")
            return
        }
        let name = FileNamer.fileName(for: Date(), ext: isPNG ? "png" : "jpg")
        do {
            // The chosen folder may have been deleted/renamed since it was set.
            try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
            try data.write(to: dir.appendingPathComponent(name))
        } catch {
            NSLog("Save failed: \(error)")
            hud.show("Couldn't save screenshot", symbol: "exclamationmark.triangle")
        }
    }

    private func ensurePermission() -> Bool {
        if PermissionManager.hasScreenRecordingPermission { return true }
        presentSetup?()   // one-button setup window owns the whole grant flow
        return false
    }

    /// Remembers who had focus before the selection overlay activates us.
    /// Recording ourselves is skipped two different ways: while our own
    /// overlay is already up (a second capture hotkey — keep the real target),
    /// and when the capture was started from one of our own windows, where
    /// there is nothing to hand focus back to.
    private func rememberFrontmostApp() {
        guard let front = NSWorkspace.shared.frontmostApplication else { return }
        guard front.bundleIdentifier != Bundle.main.bundleIdentifier else {
            if !overlay.isPresenting { previousApp = nil }
            return
        }
        previousApp = front
    }

    /// Hands focus back to that app. The remembered app is deliberately *not*
    /// cleared: re-activating an already-active app is a no-op, and keeping it
    /// means a second capture hotkey pressed while our own selection overlay
    /// is already up still has somewhere to return to.
    private func restoreFrontmostApp() {
        guard let app = previousApp else { return }
        guard FocusRestore.shouldRestore(previousBundleID: app.bundleIdentifier,
                                         ownBundleID: Bundle.main.bundleIdentifier) else { return }
        app.activate()
    }

    private func frontmostWindowID() async -> CGWindowID? {
        let content = try? await SCShareableContent.excludingDesktopWindows(
            false, onScreenWindowsOnly: true)
        return content?.windows.first(where: { $0.isOnScreen && $0.title?.isEmpty == false })?.windowID
    }
}
